### ISSUE TITLE
[GJEP, experimental] Conversion to ADEV language for generative functions

### ISSUE THREAD

**Issue by femtomc** [genjax-community/genjax#259]
In this issue, I'll propose a new interface + a program transformation which allows conversion of generative functions to probabilistic computations compatible with ADEV. I'll also discuss how to parametrize gradient strategies using ADEV by using `harvest` -- the stateful insert/pull transformation from `Oryx`.

## Gen vs. probabilistic computations

Gen explicitly traces random choices, but each generative function has two mathematical objects attached to its meaning:
* The measure $P(\tau, r; x)$ over traces and untraced randomness (if present, below I'll assume no), parametrized by arguments to the generative function.
* The deterministic return value function $f(\tau, r; x)$.

Defining a measure using a function-like language like the `Builtin` generative language often involves tracing random choices. 

A probabilistic computation in ADEV may not trace random choices (the interfaces of probabilistic computations in ADEV don't need to return choice maps, etc -- so tracing a random choice is not required). Nonetheless, there's still an underlying measure $P(\tau; x, \theta)$ (assume that we're excluding the possibility of untraced randomness). Here, $\theta$ denotes a learnable parameter. A probabilistic computation also defines a deterministic return value function $f(\tau; x, \theta)$.

In the semantics of ADEV, the expected value of the probabilistic program defines a loss function

$$
L(\theta) = E_{x \sim \[t\](\theta)}[x]
$$

* `simulate` from the probabilistic computation returns an $x$ sampled from $\[t\](\theta)$ (where $\[t\]$ here denotes the pushforward of the measure $P$ with $f$, closed over non-differentiable arguments $x$)
* `grad_estimate` of the probabilistic computation returns a gradient estimate whose expected value is the derivative of the loss (it's an unbiased estimate of the true gradient).

## Reference language (`ADEV.jl`)

[ADEV.jl](https://github.com/probcomp/ADEV.jl) defines a reference language to consider, which (combined with the [original paper](https://arxiv.org/pdf/2212.06386.pdf)) provides a set of primitives which we can implement in JAX.

```julia
function my_program(theta)
    @adev begin
        let x = @sample(normal_reparam(theta, 1)),
            b = @sample(flip_mvd(sigmoid(theta)))

            if b
                @sample(normal_reparam(x, 1))
            else
                # Recurse
                theta * @sample(my_program(theta - 1))
            end
        end
    end
end
```

### Underlying implementation

Under the hood, a CPS transformation is applied to the block - allowing propagation of dual numbers, and the ability for primitive gradient strategies (like `normal_reparam`) to implement estimators compositionally (c.f. paper).

## JAX implementation

Directly implementing ADEV is possible - JAX allows us to construct CPS-style interpreters, and propagate dual numbers. 

> NOTE: There's some challenging questions related to CPS transformation for primitives like `vmap`. Explore elsewhere.

We can define new `jax.core.Primitive` operations - here `sample`, and define primitive datatypes like `normal_reparam`, etc. This is relatively straightforward. This is what I've started here: https://github.com/probcomp/adev_jax

### GenJAX proposal

In GenJAX, I've been considering two primitives which allow `sample` to apply to generative functions, and gradient strategies to be embedded as state in pure functions (and thus, can be configured by the user externally -- using a program transformation).

```python
####################
# Sample intrinsic #
####################

sample_p = primitives.InitialStylePrimitive("sample")


def _abstract_prob_comp_call(prob_comp, *args):
    return prob_comp.simulate(*args)


def _sample(prob_comp, strat, key, args, **kwargs):
    return primitives.initial_style_bind(sample_p)(_abstract_prob_comp_call)(
        prob_comp, key, strat, args, **kwargs
    )
```

Here, `sample_p` follows the semantics of `ADEV.jl` - it requires a `prob_comp :: ProbabilisticComputation`, but it also requires a gradient strategy `strat` - I'll discuss the gradient strategy interface below.

#### Gradient strategies

The gradient strategy interface needs more work. There's a version of gradient strategies which expose them as higher-order functions which accept distributions - and then transform them to use the same density evaluation -- but a modified `grad_logpdf`. I'm pasting in the relevant appendix section from ADEV:

<img width="755" alt="image" src="https://user-images.githubusercontent.com/34410036/222250816-28b0f6ce-b5bc-4ded-9a74-e616cac4f742.png">

Starting slow, the ADEV primitive interface that users will actually use is `sample`:

```python
@typecheck
def sample(
    prob_comp: AbstractProbabilisticComputation, key: PRNGKey, args: Tuple, **kwargs
):
    # Default strategy is REINFORCE.
    strategy = strat(Reinforce(), "sample")
    return _sample(prob_comp, key, strategy, args, **kwargs)
```

Here, `sample` calls another primitive-wrapping function `strat`. `strat` embeds a gradient strategy into the program -- using a stateful transformation interface called `harvest`. The default strategy is `REINFORCE`.

```python
NAMESPACE = "adev_strategy"
adev_tag = functools.partial(harvest.sow, tag=NAMESPACE)


@typecheck
def strat(strategy: GradientStrategy, addr):
    return adev_tag(strategy, name=addr)
```

Here, `harvest.sow` "sows" state into a computation - and the harvest transformation lets you insert/pull this state out -- allowing a user to reconfigure the gradient strategy corresponding to specific `sample` sites.

## Naive conversion

> TODO: not clear if this design is the right one yet. After discussing with Alex -- probably better to explicitly implement `ProbabilisticPrimitive` -- and provide interfaces that support the implementation of gradient strategies specialized to each primitive (as in the paper).

In GenJAX, I've been considering interfaces that allow generative functions to be transformed into the ADEV language -- these include an interface called `adev_convert` which implements the forward sampling process in terms of ADEV primitives. Generative functions can opt into this interface - the `Distribution` language supports `adev_convert` implementations that directly use `sample` -- whereas the `Builtin` language supports a program transformation which defers `adev_convert` to callees, etc.

This doesn't seem like the only way we can do this -- and, indeed, for black box generative functions -- we'll likely want a way to define new "leaf" primitives under the ADEV computation interface, with built in `grad_logpdf`. Implementing `grad_logpdf` allows usage of estimators like REINFORCE (c.f. discussion above).

```python
#######
# ADEV #
#######

def adev_convert(self, key: PRNGKey, args: Tuple) -> Tuple[PRNGKey, Any]:
    """Convert a generative function to a forward sampling function which
    uses primitives which are compatible with ADEV's probabilistic
    computation language."""
    raise NotImplementedError
```

Here's an example of `adev_convert`:

```python
@typecheck
def adev_convert(self, key: PRNGKey, args: Tuple):
    prob_prim = adev.ProbabilisticPrimitive(self)
    key, v = adev.sample(prob_prim, key, args)
    return key, v
```
